libs/MeshKernel/src/Mesh2D.cpp

```diff
@@ -1934,3 +1934,645 @@ std::vector<int> Mesh2D::NodeMaskFromPolygon(const Polygons& polygon, bool insid
+{
+    //
+    // 2------4------+------+
+    // |      |      |      |
+    // |      |      x------+
+    // |      o      |      |
+    // |      |      x------+
+    // |      |      |      |
+    // 1------3------+------+
+    //
+
+    // Compute point labeled with 'o' in ASCII diagram above
+    Point midPoint = PointAlongLine(m_nodes[startNode], m_nodes[endNode], 0.5);
+    UInt newNodeIndex = InsertNode(midPoint);
+
+    // Connect node marked with 'x' to nodes labeled 3 and 'o'
+    ConnectNodes(hangingNodes[0], newNodeIndex);
+    ConnectNodes(hangingNodes[0], startNode);
+
+    // Connect node marked with 'x' to nodes labeled 'o' and 4
+    ConnectNodes(hangingNodes[1], newNodeIndex);
+    ConnectNodes(hangingNodes[1], endNode);
+
+    // Connect node marked with 'o' to nodes labeled 3 and 4
+    ConnectNodes(newNodeIndex, startNode);
+    ConnectNodes(newNodeIndex, endNode);
+
+    UInt adjacentFaceId = NextFace(faceId, edgeId);
+
+    DeleteEdge(edgeId);
+
+    if (adjacentFaceId != constants::missing::uintValue)
+    {
+        UInt nextOppositeEdge = FindOppositeEdge(adjacentFaceId, edgeId);
+        // Connect node marked with 'o' to nodes labeled 1 and 2
+        ConnectNodes(newNodeIndex, m_edges[nextOppositeEdge].first);
+        ConnectNodes(newNodeIndex, m_edges[nextOppositeEdge].second);
+    }
+}
+
+void Mesh2D::FreeThreeHangingNodes(const UInt faceId, const UInt edgeId, const HangingNodeIndexArray& hangingNodes, const UInt startNode, const UInt endNode)
+{
+    //
+    // 2------4------+------+
+    // |      |      |      |
+    // |      |      x------+
+    // |      |      |      |
+    // |      o      x------+
+    // |      |      |      |
+    // |      |      x------+
+    // |      |      |      |
+    // 1------3------+------+
+    //
+
+    // Compute point labeled with 'o' in ASCII diagram above
+    Point midPoint = PointAlongLine(m_nodes[startNode], m_nodes[endNode], 0.5);
+    UInt newNodeIndex = InsertNode(midPoint);
+
+    ConnectNodes(hangingNodes[1], newNodeIndex);
+    ConnectNodes(newNodeIndex, endNode);
+    ConnectNodes(newNodeIndex, startNode);
+
+    ConnectNodes(hangingNodes[0], newNodeIndex);
+    ConnectNodes(hangingNodes[0], startNode);
+
+    ConnectNodes(hangingNodes[2], newNodeIndex);
+    ConnectNodes(hangingNodes[2], endNode);
+
+    UInt adjacentFaceId = NextFace(faceId, edgeId);
+
+    DeleteEdge(edgeId);
+
+    if (adjacentFaceId != constants::missing::uintValue)
+    {
+        UInt nextOppositeEdge = FindOppositeEdge(adjacentFaceId, edgeId);
+        ConnectNodes(newNodeIndex, m_edges[nextOppositeEdge].first);
+        ConnectNodes(newNodeIndex, m_edges[nextOppositeEdge].second);
+    }
+}
+
+void Mesh2D::FreeFourHangingNodes(const UInt faceId, const UInt edgeId, const HangingNodeIndexArray& hangingNodes, const UInt startNode, const UInt endNode)
```

Comment:
can you make all the dd boundary algorithm a separate class?

Reply:
Done